### DISCUSSION TITLE
ExternalTaskSensor manually trigger

### DISCUSSION THREAD

**Discussion by Lee2532** [apache/airflow#24250] [General]
### Description

when the External Task Sensor is manually executed, not work

### Use case/motivation

We can add options to perform functions such as scheduling when executing manually.

Since we FAIL the DAG with External Task Sensor when executing manually, we add logic to pass when executing manually

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Could you please be more precise and explain your use case in more detail? What problem are you really trying to solve? Ideally explaining it on examples that you experienced.

**Comment by Lee2532** [comment]
When using an external task sensor, running DAG manually does not work normally.
Therefore, I would like to add a parameter that allows the corresponding external task sander to pass when triggering the corresponding TASK manually.

for example
`
wait_airflow_test = ExternalTaskSensor(
        task_id="wait_airflow_test",
        external_dag_id="airflow_sample",
        external_task_id="start",
        allowed_states=["success"],
        failed_states=None,
        execution_delta=None,
        execution_date_fn=None,
        check_existence=True,
        dag=dag,
        default_args=dag_args,
        mode='reschedule',
        timeout=5,
        wait_task=False
    )


`

**Reply by jayarjo** [reply]
Do you say that ExternalTaskSensor doesn't sense when DAG is run manually?

**Reply by Lee2532** [reply]
yes. i want ExternalTaskSensor test but manually  is not work

**Comment by potiuk** [comment]
How about [params](https://airflow.apache.org/docs/apache-airflow/stable/concepts/params.html?highlight=params)  - I still do not understand what you are asking for - but that would be my guess tha it's likely what you want. In either case I am converting that to a discussion until this is clarified.

**Reply by josh-fell** [reply]
@potiuk Should we set #24104 (related PR to implement this feature) as a draft for now? Close it? WDYT?

**Reply by potiuk** [reply]
To be honest I do not know because I am rather lost what the problem was about.

**Reply by josh-fell** [reply]
Same. I was just thinking about an explicit way to not have that PR get merged until we can understand the impetus behind it. I was leaning towards setting it to draft in the interm.

**Reply by potiuk** [reply]
I closed. We can always re-open whe @Lee2532 clarifies what is the problem that is being solved.

**Comment by jon-evergreen** [comment]
Hi @potiuk 

I can't be sure that this is exactly what @Lee2532 is talking about, but I think I'm correct (and the fix proposed in the PR would solve the problem I currently have - though I don't like how it fixes it).

We use `ExternalTaskSensor` in a few places in our airflow deployment.  The most notable place is at the start of our dbt running DAG.  The idea is to have a bunch of `ExternalTaskSensor`s for the ETL dags so that the dbt models are not run until yesterdays data have loaded. even if that ends up being late due to failures, retries, whatever.  This works great for scheduled runs.

However, if you do a manual run of the DAG, the `ExternalTaskSensor` looks for a run of the ETL dags with the appropriate execution date which is, by the `ExternalTaskSensor` logic, the current datetime, which probably doesn't exist.  It would be helpful in these circumstances to have a way of saying "Please treat these `ExternalTaskSensor` as passed" so that you can run the DAG manually.

The alternatives all seem not so great from my point of view:

1. have a second version of the DAG with no sensors (& no schedule) purely for manual triggering.  But then that's two dags to keep in sync, and two places to look for history
2. Have this little widget I developed from a few different tasks which looks roughly like:

```python
  def wait_if_scheduled(dag_run, ti: TaskInstance):
    """
    Determines whether to run the wait or the no wait task based on if the run is scheduled or not
    """
    group_name = ti.task_id.split(".", 2)[0]
    if dag_run.external_trigger is True:
        return f"{group_name}.waited"
    else:
        return f"{group_name}.wait"

  with TaskGroup("external_dag_waiter") as elt_wait:
      branch = BranchPythonOperator(
          task_id="maybe_skip", python_callable=wait_if_scheduled
      )
      waiter = ExternalTaskSensor(
          task_id="wait",
          external_dag_id="external_dag",
          poke_interval=300,
          mode="reschedule",
      )
      waited = EmptyOperator(
          task_id="waited", trigger_rule="none_failed_min_one_success"
      )

      branch >> waiter >> waited
      branch >> waited
```
But that's quite a bit of noise in a DAG.

**Reply by potiuk** [reply]
I think you should use 2.4's Dataset Driven Scheduling to achieve what you want. External Task Sensors are really poor way of managing dependencies and using `micropiplines" and Dataset Driven scheduling is WAY better and future proof.
